python/ray/autoscaler/v2/instance_manager/reconciler.py

```diff
@@ -163,3 +191,237 @@ def _step_next(
+            )
+            if not update_event:
+                continue
+
+            logger.debug("Updating {}".format(MessageToDict(update_event)))
+            updates[instance.instance_id] = update_event
+
+        # Update the instance manager for the events.
+        Reconciler._update_instance_manager(instance_manager, updates, version)
+
+    @staticmethod
+    def _try_or_fail_allocation(
+        im_instance: IMInstance,
+        unassigned_cloud_instances_by_type: Dict[str, List[CloudInstance]],
+        launch_errors: Dict[str, LaunchNodeError],
+    ) -> Optional[IMInstanceUpdateEvent]:
+        """
+        Allocate, or fail the cloud instance allocation for the instance.
+
+        Args:
+            im_instance: The instance to allocate or fail.
+            unassigned_cloud_instances_by_type: The unassigned cloud instances by type.
+            launch_errors: The launch errors from the cloud provider.
+
+        Returns:
+            Instance update to ALLOCATED: if there's a matching unassigned cloud
+                instance with the same type.
+            Instance update to ALLOCATION_FAILED: if the instance allocation failed
+                with errors.
+            None: if there's no update.
+
+        """
+        unassigned_cloud_instance = None
+
+        # Try to allocate an unassigned cloud instance.
+        # TODO(rickyx): We could also look at the launch request id
+        # on the cloud node and the im instance later once all node providers
+        # support request id. For now, we only look at the instance type.
+        if len(unassigned_cloud_instances_by_type.get(im_instance.instance_type, [])):
+            unassigned_cloud_instance = unassigned_cloud_instances_by_type[
+                im_instance.instance_type
+            ].pop()
+
+        if unassigned_cloud_instance:
+            return IMInstanceUpdateEvent(
+                instance_id=im_instance.instance_id,
+                new_instance_status=IMInstance.ALLOCATED,
+                cloud_instance_id=unassigned_cloud_instance.cloud_instance_id,
+            )
+
+        # If there's a launch error, transition to ALLOCATION_FAILED.
+        launch_error = launch_errors.get(im_instance.launch_request_id)
+        if launch_error and launch_error.node_type == im_instance.instance_type:
+            return IMInstanceUpdateEvent(
+                instance_id=im_instance.instance_id,
+                new_instance_status=IMInstance.ALLOCATION_FAILED,
+                details=launch_error.details,
+            )
+        # No update.
+        return None
+
+    @staticmethod
+    def _handle_cloud_instance_terminated(
+        instance_manager: InstanceManager,
+        non_terminated_cloud_instances: Dict[CloudInstanceId, CloudInstance],
+    ):
+        """
+        For any IM (instance manager) instance with a cloud node id, if the mapped
+        cloud instance is no longer running, transition the instance to TERMINATED.
```

Comment:
hope the cloud provider is at least sequential consistent.
if it turns out our expectation is too high, we might give those disappeared node a grace period

Reply:
good point - let me add a comment to the cloud provider.